geocoding_android/android/src/main/java/com/baseflow/geocoding/Geocoding.java

```diff
@@ -34,26 +43,68 @@ void setLocaleIdentifier(@Nullable Locale locale) {
      * @return a list of Address objects. Returns null or empty list if no matches were found or there is no backend service available.
      * @throws java.io.IOException if the network is unavailable or any other I/O problem occurs.
      */
-    List<Address> placemarkFromAddress(String address) throws IOException {
-
+    @SuppressWarnings("deprecation")
+    void placemarkFromAddress(String address, GeocodeListenerAdapter callback) throws IOException {
         final Geocoder geocoder = createGeocoder(androidContext, locale);
-        return geocoder.getFromLocationName(address, 5);
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            getAddressesWithGeocodeListener(geocoder, address, 5, callback);
+        } else {
+            List<Address> addresses = geocoder.getFromLocationName(address, 5);
+            callback.onGeocode(addresses);
+        }
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.TIRAMISU)
+    private void getAddressesWithGeocodeListener(Geocoder geocoder, String address, int maxResults, GeocodeListenerAdapter callback) {
+        geocoder.getFromLocationName(address, maxResults, new Geocoder.GeocodeListener() {
+            @Override
+            public void onGeocode(List<Address> geocodedAddresses) {
+                callback.onGeocode(geocodedAddresses);
+            }
+
+            @Override
+            public void onError(@Nullable String errorMessage) {
+                callback.onError(errorMessage);
+            }
+        });
     }
 
     /**
      * Returns a list of Address objects matching the supplied coordinates.
      *
-     * @param latitude the latitude point for the search
+     * @param latitude  the latitude point for the search
      * @param longitude the longitude point for the search
      * @return a list of Address objects. Returns null or empty list if no matches were found or there is no backend service available.
      * @throws IOException if the network is unavailable or any other I/O problem occurs.
      */
-    List<Address> placemarkFromCoordinates(
+    @SuppressWarnings("deprecation")
+    void placemarkFromCoordinates(
             double latitude,
-            double longitude
+            double longitude,
+            GeocodeListenerAdapter callback
     ) throws IOException {
         final Geocoder geocoder = createGeocoder(androidContext, locale);
-        return geocoder.getFromLocation(latitude, longitude, 5);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            getLocationWithGeocodeListener(geocoder, latitude, longitude, 5, callback);
+        } else {
+            callback.onGeocode(geocoder.getFromLocation(latitude, longitude, 5));
```

Comment:
Similar here, extract only the deprecated code into a separate method and suppress warnings on that method alone.